[package] fix whichmodule for OrderedImporter

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#55646 [package] fix whichmodule for OrderedImporter**



Differential Revision: [D27664238](https://our.internmc.facebook.com/intern/diff/D27664238)

suo committed:
[package] fix whichmodule for OrderedImporter

[ghstack-poisoned]

SplitInfinity commented on torch/package/importer.py:
```diff
@@ -184,3 +184,11 @@ def import_module(self, module_name: str) -> ModuleType:
+
+    def whichmodule(self, obj: Any, name: str) -> str:
+        for importer in self._importers:
+            module_name = importer.whichmodule(obj, name)
```
Might the importer implementation raise `ModuleNotFoundError`?

suo replied:
`whichmodule` is not supposed to actually import any modules or have side effects into the import system, so I don't think so.

SplitInfinity commented on torch/package/importer.py:
```diff
@@ -184,3 +184,11 @@ def import_module(self, module_name: str) -> ModuleType:
+            if module_name != "__main__":
+                return module_name
+
+        return "__main__"
```
This is what the CPython `whichmodule` does too but I don't really understand why. 😕  Is it understood to indicate that a module wasn't found?

suo replied:
honestly I am not sure, I suppose the idea is that an object without `__module__` defined that also isn't accessible from any known module *must* have been defined interactively/from a script.

SplitInfinity commented on test/package/test_importer.py:
```diff
@@ -84,6 +85,32 @@ def test_ordered_importer_basic(self):
+        self.assertEqual(foo_then_bar.whichmodule(DummyClass(), ""), "foo")
+
+        bar_then_foo = OrderedImporter(dummy_importer_bar, dummy_importer_foo)
+        self.assertEqual(bar_then_foo.whichmodule(DummyClass(), ""), "bar")
```
Consider adding a case in which the first importer fails to find the module that the object belongs to.

suo committed:
Update on "[package] fix whichmodule for OrderedImporter"




Differential Revision: [D27664238](https://our.internmc.facebook.com/intern/diff/D27664238)

[ghstack-poisoned]
